### DISCUSSION TITLE
Confusion about sleep lights in 3.0

### DISCUSSION THREAD

**Discussion by ReplicaObscura** [jseidl/magic-areas#247] [Q&A] [answered]
I'm not sure what I am missing in version 3 to get my sleep lights working like before.

In most of my rooms, I want only one or two of the room lights to turn on when motion is detected during sleep, while in other rooms (like the bedroom) I want no lights on while sleeping.

I think I configured this right by not selecting any sleep lights in the bedroom, and selecting the lights I want to still be controlled via motion at night for the other rooms. I have the sleep entity state set to "on" in the settings, and I have selected my Adaptive Lighting sleep mode binary sensor.

I also am using the "occupied" state to control the lights during sleep mode for the rooms I still want lights in at night, so they should work just like during the day. However, when I turn Adaptive Lighting's sleep mode on, all my lights stop being controlled regardless of my settings.

Is there some kind of guide on how to get sleep lights working right in version 3.0, or am I missing something obvious?

**Comment by jseidl** [comment]
In 3.0 the light group names are merely descriptive, the won't automatically be tied to any state.  In order for the 'Sleep Lights' group turn on when sleep state is on, you have to add the "sleep" state to them.

Check the [wiki for this feature](https://github.com/jseidl/hass-magic_areas/wiki/Light-Groups) for more information on that. Let me know if I misunderstood your question.

**Reply by ReplicaObscura** [reply]
I thought that selecting "sleep" there would mean the sleep lights are on the entire time sleep mode is on, like a night light? If that's correct, then that's not what I'm looking for, but I might be misunderstanding the documentation so I'll give it a try.

What I'm looking for is for my light automations to still work on sleep mode, but only for the lights I've specified as sleep lights.

**Reply by jseidl** [reply, marked as answer]
So the logic is quite involved and having who wrote the code, write the documentation causes these issues 😆 

Taking it from the top, some axioms

1. Area light control, it the highest level, is tied with area occupancy. Meaning area occupied -> logic will run to turn on, area clear -> lights will turn off (if configured). 
2. Area light control depends on the switch `switch.area_light_control_$area_name` to be on

Then the flow is as follows
1. If all the above constraints are met (area is occupied and area_light_control switch is on, then proceed
2. If there's a dark entity configured, lights will only turn on if dark, if not they'll turn on every time the room is occupied
3. When an area is occupied, it will fire an `AREA_STATE_CHANGED` event, Light Groups, listening to this event, will check for the area states
4. Area states are divided into primary (occupied/clear) and secondary (dark/bright, extended, sleep, accented) (see https://github.com/jseidl/hass-magic_areas/wiki/Area-State)
5. If area changed from occupied -> clear, turn off all lights
6. If area changed from clear to occupied, check priority states (sleep, accented). If any of those states are present, turn the Light Groups that are tied to these states
7. If no priority states are on, check if the Light Group is tied to the `occupied` state. If none, don't turn any lights on

The `extended` state will never be present at first, but only after the configured seconds passed. When this state pops up, the same routine will happen, checking if there's any light group tied to this state.

If you want lights to be present in multiple states (i.e. not be turned off when changing, say, from occupied to sleep) you have to add it to both states.

I'll give a few examples of my lights so you can get an idea

> Axiom, I use binary_sensor.area_light_exterior as my dark sensor, I use adaptive lighting's sleep switch as sleep entity for all areas. At midnight an automation flips all sleep switches on, at 6am turns them off

### Living room
- I have my floor lamp as a Task Light, and I have the Task Lights group listen to the `occupied` and `extended` states
- I have my ceiling lights as Overhead Lights, listening to the `extended` state (I only want them to turn on if I'm "staying" at the living room, if I'm only passing by, the floor lamp is alright)
- I have two Philips Hue Play on the floor as Accent Lights, those set to the `accented` and `sleep` state. I have the TV being on (on a template sensor) that is my accented entity. This causes all lights to go off except the Accent Lights. Also if I walk through the room during sleep hours only those Accent Lights come up

### Bathroom
- This don't have a dark entity, as this room has no windows, it's always dark there

### Entrance hall
- I have the single overhead light in the Overhead Lights group, but this group is not tied to the `occupied` state as I don't generally want that light on when I'm passing by
- This Overhead lights group is actually tied to the `extended` and `accented` states, and I use the front door as an accented state. This means lights will only turn on when I open the door or if I hang out in the entrance hall long enough

### Kitchen & Family room
- Pretty straightforward, I have overhead lights tied to `occupied` and `extended` states. I have a Hue Go (kitchen) and a Nanoleaf (family room) as Accent Lights, tied to the `sleep` state, so during sleep hours only those come in, otherwise the ceiling lights come in.


Hope those examples help. Remember `occupied` state is not "all mighty" meaning if you want those lights to stay on, you'll have to add it to the other states as well, otherwise when you move into `extended` (for example) they will be turned off. 

Lastly these states can be tracked only when the area goes from clear to occupied or whenever these secondary states change (or both), this is controlled by the option that has `occupancy` / `state` on them, because sometimes you don't want the lights to change if you're already there, only when first entering the room.

Lemme know if you need additional guidance on this!

**Reply by ReplicaObscura** [reply]
Thank you so much! Reading through this has been very insightful, and I think I can understand how to do most of the things I've been wanting to do with my lights based on that info.

As someone relatively new to MA (I started actually using it a few weeks before v3 came out), there were both things I didn't understand fully from v2, and then different things that I didn't understand after transitioning to v3.

Something that really tripped me up is that it seems like states operate on several levels, and interact with each other and with the light groups, in ways that weren't fully apparent, or at least intuitive, to me previously. I think I have a much better grasp on this now, so I'm going to try to configure everything the way I want and see how it goes!

I have a couple of other PRs to work on related to recent discussions anyway, so I could possibly help refine some of the documentation a bit from the perspective of someone who has been recently learning all of these concepts and the pitfalls I've encountered.

**Reply by jseidl** [reply]
Glad that helped! Yeah the state based stuff is pretty involved and I've done my best on the wiki but I have several concepts on my mind that I take for granted and wasn't very successful on explaining in the wiki.

2x had the feature listen to entities directly and have a bunch of state logic inside. With 3x I've moved state tracking to the main entity and made it fire events that the features could listen on to simplify feature building but indeed with that comes conventions and other stuff that aren't very straightforward. Those changes were made to make it less cemented and more async ready with the event based approach. 

I'd appreciate a lot if you're able to enhance the wiki from a newcomer point of view. I have the plan to make a video explaining all the features but had a baby recently and time is of a premium.

Lemme know if you have further questions!
